### DISCUSSION TITLE
Can we do changing colors in json file of lottie animation and return "Animation"? (iOS, Swift)

### DISCUSSION THREAD

**Discussion by pr0khorovmaksim** [airbnb/lottie-ios#1720] [Ideas]
Dear Lottie Team and developers

Now we change the colors of the lottie animation as follows:

let animationView = AnimationView()
let animationView.setValueProvider(colorProvider, keypath: keypath)

The ideal solution would be to prepare animations in advance by changing their colors in JSON

It is necessary to prepare animations with other colors in advance, the current implementation does not allow this. I need to get a ready Animation or JSON output so that I can then assign it to AnimationView

need the following functions, example:

func updateJSON() -> Animation
func updateJSON() -> JSON (update json file)

Thank you!
Max

**Comment by calda** [comment]
https://lottiefiles.com has an editor that lets you modify the colors of layers and then re-export an updated json file. It's really useful! 

I think visual editors like that (or modifying the colors in the original after effects file) are likely a better UX than making modifications using Swift code.

**Reply by pr0khorovmaksim** [reply]
@calda, Thank you for your reply!

Yes, I know that there is such a possibility. But we need a software solution, I'm sure our lottie community needs it too.

We have 20 animations in the project, each animation can be in two variations: either a black silhouette or a color animation. And it turns out 40 animations. Then we plan to constantly add animations, since we actively decided to work with them and this is very important for us. In order not to paint each animation separately through AE or through lottiefiles, or through animationView.setValueProvider(colorProvider, keypath: keypath), it would be nice if we could prepare and generate the animations themselves through the code.

**Reply by calda** [reply]
Could you share more about why `animationView.setValueProvider(colorProvider, keypath: keypath)` is unsuitable for your use case?

**Reply by pr0khorovmaksim** [reply]
@calda, I made a loading screen in the application, it prepares all the lottie animations, there I use
let animation = Animation.named("...") and I get 20 "Animation", the animation preparation time on the next screen is critical for us, because if we use setValueProvider on the next screen - we paint over all layers in the animation in black, on it takes a while and this animation starts to lag behind the others (they are all the same for testing purposes). I tried to do this on the loading screen: let animation = Animation.named("...") -> animationView.setValueProvider(colorProvider, keypath: keypath) -> let animation = animationView.animation. I thought that I could take the shaded animation in this way and be able to solve my problem with it, alas, this is not how it works. I searched the entire community - I did not find an elegant solution in the preparation of animations ..

**Reply by calda** [reply]
This seems like it could potentially be a bug, would you be interested in creating / sharing a sample project that demonstrates this?

**Reply by pr0khorovmaksim** [reply]
@calda, I don't know if I can do it, I'll think about it tomorrow.

I will write here briefly how it works. We have a collectionView, in each cell there is a lottie animation for the entire size. Cells in the collection can be 10 for example. When we click on the cell, we have to paint the lottie animation black, if we click again, the color returns to the original one (The lottie library does not provide for the return of animation with changed colors to the original animation that was before the changes, but this is not so important, I am writing so that you are aware, all of a sudden you add new functionality). So, when there are a lot of animations, problems appear, which I voiced above. 
Lotty Animation that we blacked out and then re-colored - it will lag behind the others. And in order to avoid such a problem, I need to programmatically prepare a lottie animation in advance in black, today I wanted to parse Animation.json so that there in "k": (...,...,...) to set black color is difficult to do from scratch, since we already have a ready-made lottie library - that's why I wrote to you. I'm also using the "renderingEngine = .coreAnimation" you recently added - it's a huge performance improvement thank you!
